packages/shared/components/FileTransfer/FileTransfer.test.tsx

```diff
@@ -0,0 +1,173 @@
+
+import { FileTransfer, TransferHandlers } from './FileTransfer';
+import { FileTransferContextProvider } from './FileTransferContextProvider';
+import {
+  FileTransferDialogDirection,
+  FileTransferListeners,
+} from './FileTransferStateless';
+
+test('click opens correct dialog', () => {
+  render(
+    <FileTransferContextProvider
+      openedDialog={FileTransferDialogDirection.Download}
+    >
+      <FileTransfer beforeClose={undefined} transferHandlers={undefined} />
+    </FileTransferContextProvider>
+  );
+  expect(screen.getByText('Download Files')).toBeInTheDocument();
+});
+
+test('downloads component changes when file transfer callbacks are called', async () => {
+  let listenersMock: FileTransferListeners = {
+    onComplete(): void {},
+    onError(): void {},
+    onProgress(): void {},
+  };
+
+  const handler: TransferHandlers = {
+    getDownloader: async () => fileTransferListeners => {
+      listenersMock = fileTransferListeners;
+    },
+    getUploader: async () => undefined,
+  };
+  render(
+    <FileTransferContextProvider
+      openedDialog={FileTransferDialogDirection.Download}
+    >
+      <FileTransfer beforeClose={undefined} transferHandlers={handler} />
+    </FileTransferContextProvider>
+  );
+  fireEvent.change(screen.getByLabelText('File Path'), {
+    target: { value: '/Users/g/file.txt' },
+  });
+  fireEvent.click(screen.getByText('Download'));
+  const listItem = await screen.findByRole('listitem');
+  expect(listItem).toHaveTextContent('/Users/g/file.txt');
+
+  act(() => listenersMock.onProgress(50));
+  expect(listItem).toHaveTextContent('50%');
+
+  act(() => listenersMock.onComplete());
+  expect(listItem).toContainElement(screen.getByTitle('Transfer completed'));
+
+  act(() => listenersMock.onError(new Error('Network error')));
+  expect(listItem).toHaveTextContent('Network error');
+});
+
+test('onAbort is called when user cancels upload', async () => {
+  let abortControllerMock: AbortController;
+
+  const handler: TransferHandlers = {
+    getDownloader: async () => (fileTransferListeners, abortController) => {
+      abortControllerMock = abortController;
+    },
+    getUploader: async () => undefined,
+  };
+  render(
+    <FileTransferContextProvider
+      openedDialog={FileTransferDialogDirection.Download}
+    >
+      <FileTransfer beforeClose={undefined} transferHandlers={handler} />
+    </FileTransferContextProvider>
+  );
+  fireEvent.change(screen.getByLabelText('File Path'), {
+    target: { value: '/Users/g/file.txt' },
+  });
+  fireEvent.click(screen.getByText('Download'));
+  fireEvent.click(await screen.findByTitle('Cancel'));
+  expect(abortControllerMock.signal.aborted).toBeTruthy();
+});
+
+test('file is not added when  transferHandler does not return anything', async () => {
+  const handler: TransferHandlers = {
+    getDownloader: async () => undefined,
+    getUploader: async () => undefined,
```

Comment:
The handler returning nothing is my way to handle a situation where user [cancels selecting the path for a file to download](https://github.com/gravitational/webapps/blob/aa8afe81f2bad2b32646128e3e03c20397fb5d92/packages/teleterm/src/ui/DocumentTerminal/DocumentTerminal.tsx#L78). If nothing is returned I skip adding file to the list. We could theoretically throw an error, do you think this will make it easier to use the API?

Reply:
Done